README.md
[![Build](https://github.com/eliasku/13/actions/workflows/build.yml/badge.svg)](https://github.com/eliasku/13/actions/workflows/build.yml)
[![Twitter](https://img.shields.io/twitter/follow/eliaskuvoice.svg?style=flat&label=Follow&logoColor=white&color=1da1f2&logo=twitter)](https://twitter.com/eliaskuvoice)
[![Play Now](https://img.shields.io/badge/Play-online-pink.svg)](https://iioi.herokuapp.com/)

# ☠️ 13 ☠️

Fast-paced multiplayer top-down shooting game for [js13k 2022 competition](https://js13kgames.com/)

[🚪PLAY NOW](https://next13.herokuapp.com/)

<p align="center">
<img src="13.gif">
</p>

- 🎖️ **7-th place** winner!
- 🥇 **1-st place** in **Server** category
- 🥉 **3-rd place** in **Mobile** category
- Check out [13 game post-mortem](https://eliasku.hashnode.dev/13-game)
- Check out initial 13`kb` submission source-code in [`js13k` branch](https://github.com/eliasku/13/tree/js13k)

## 📖 How to play

Start game, share the link and wait friends to play:

- The latest release is available on [🚪iioi](https://iioi.herokuapp.com/) server
- [Online Players stats](https://eliasku-games.web.app/13/servers.html)

### 🎮 Controls

#### 🖱️Mouse | ⌨️ Keyboard

- Move Mouse to aim and look around
- Push Mouse button to shoot
- Use `W A S D` / `↑ ← ↓ →` to move
- Hold `Shift` to slow down moving
- Press `E` to to PICK weapon on the map or DROP your current weapon
- Press `Space` to jump
- Press `R` to reload weapon
- Press `Q` to switch secondary weapon slot

#### 📱 Touch-screen

- Use Left Joystick to walk-run-jump
- Use Right Joystick to aim-shoot (Move around aim-shoot zones to trigger some weapons like Pistol)
- Use `DROP` button to pick weapon on the map or drop your current weapon
- Use `RELOAD` button to reload the weapon
- Use `SWAP` button to swap the secondary weapon slot

### 📜 Rules

At spawn each Player or NPC has 10`hp` (hit points). You are able to get extra 10`sp` (shield points) on top of that.

- Player kills neutral NPC to gain +1 score.
- Player kills another Player to gain +10 scores.

Kill opponents 💀 | 👹 | 🤡 | 🤖 | 🎃 | 🦝 | 🐙 | 🐰 | 🦌 | 🐺 | 🐵 | 🦊 | 🐭 | 🦍 | 🐸 to score `FRAG` and get 5`cr`

Kill NPC 🍅 | 😐 | 🐷 | 🎅🏻 to score `FRAG` and get 1`cr`

### Items

Destroy objects 🛢 | 📦 | 🪦 to get items

- ❤️ Heart: pick to restore 1`hp`
- 💊️ Pill: pick to restore 2`hp`
- 🪙️ Coin: pick for 1`cr` (credits)
- 💎️ Diamond: pick for 5`cr` (credits)
- 🛡️ Shield: pick to add 1`sp` (shield-point)
- 🧱️ Ammo Magazine: pick to add 1`am` (ammo-magazine)

### Weapons

- 🔪 Knife (melee)
- 🪓 Axe (melee)
- 🔫 Pistol (trigger)
- 🖊 Machine-gun (auto)
- ️✏️ Heavy machine-gun (auto)
- 🪥 Shotgun (bouncing, scatter)
- ⛏ Crossbow (high velocity)
- 🔌 Plasma-gun (bouncing, auto)
- 🧵 Rail-gun (piercing)
- 🧣 Uzi (tracer bullets, rapid fire)

### Dev-menu

Tap 4 times on the main game logo to unlock dev-mode and dev-settings.

## Minimal requirements

### Client

- **Good, low-latency network connection** is required for each playing client
- **Fast mobile device or desktop** to not lag other clients
- WebAudio `AudioContext` support is required (available from Safari iOS 14.5, April 2021)
- WebGL context is required
- Modern JS syntax support
- Checked in the latest Chrome, Safari, Firefox on iOS, Android and Mac.

### Server, build stack

- `NodeJS` **v18 or higher** is required
- `NPM` **v7 or higher** is required for workspaces

## 📦 Resources

Code for Music generation in runtime created by [author](https://twitter.com/eliaskuvoice). Some instrument samples are picked from [ZzFXM](https://keithclark.github.io/ZzFXM/) example song `Depp`

Emoji Font [Twemoji Mozilla](https://github.com/mozilla/twemoji-colr/releases) is used for cross-platform emoji rendering. Game is able to work without `e.ttf` file, but some icons are incorrect rotation angle, or different at all. But game should be playable anyway.

2D graphics rendering is started from [js13k-2d](https://github.com/kutuluk/js13k-2d) and highly rewritten for what I need.

Sound Effects - [ZZFX](https://github.com/KilledByAPixel/ZzFX).

## How to build

First of all, install dependencies:
```shell
npm i
```

Use `start` script to build and watch for changes, run local server ([localhost:8080](http://localhost:8080))

```shell 
npm start
```

To deploy the game run the `build` script and then start the server
```shell
# build only
npm run build
# run server
node server.js
```

### Modify game config

1. Recommended way to change global game configuration is via `packages/tools/src/config.ts` file. It provides types and some good utilities to compose the final configuration.
2. Generate `packages/client/assets/config.json` by running `npm run config`.
3. Run `npm start` or `npm run build` - scripts will copy `packages/client/assets/config.json` to `public/config.json`, it will be loaded by the game and will be used for all global configuration.

If you need some new option from the code to be exposed in `config.json`, please create the issue.


CHANGELOG.md
## 1.0.26

- Fix arrow keys scrolls parent document
- Add language selection
- Change `sp`/`hp`/`cr`/`mags` to icons

## 1.0.25

- Antialiasing option
- Map generation
- Damage hit labels
- Critical hit chance
- Respawn screen
- High quality text rendering
- Support multiple client versions on server
- Fix key controls bubble to iframe parent
- Simple modal popups instead of browser's prompts

## 1.0.24

- Turn off git info hash from server version verify

## 1.0.23

- Re-enable TURN servers

## 1.0.22

- Remove analytics from Poki build

## 1.0.21

- Move pages to site
- Add firebase analytics  

## 1.0.19

- Add new Uzi weapon 🧣 (tracer bullets)
- Add Laser sight and enable for all guns!
- Walls and map boundaries stops Rail-gun ray projectile
- Load game configuration from JSON file #21
- Fix camera feedback on Y-axis

## 1.0.18

- Add draft walls
- Enable High-DPI by default
- More padding for dynamic font
- Autoplay continue attacks if low-hp and no hp items
- Rework room join flow
- Add top-state out of sync disconnect for Production mode

## 1.0.17

- Add screen shake on object death
- Improve weapon camera feedback
- Fix missing Frog avatar emoji

## 1.0.16

- Fix replays
- Add copy game link URL
- Fix join by game-code
- Add support for checking online on different servers

## 1.0.15

- (dev) `watch` script renamed to `start`
- Fix sound and music

## 1.0.14

- Add simple autoplay mode for debugging
- Change user settings data
- Add replay files initial support (wip)
- Player bot API (wip)

## 1.0.13

- Add in-game menu (graceful way to quit the game room)
- Fix rare font atlas artifacts
- Change settings button in Main Menu
- Add settings for Practice Mode 

## 1.0.12

- Optimize and refactoring game state data
- Changes for [poki](https://poki.com/)

## 1.0.11

- Create and join game rooms #38
- New FROGMAN head
- Brand icons
- RTC packet lag calculation #73

## 1.0.10

- Fix render occlusion
- Fix desync item's lifetime overflow
- Fix start tic sync issue (disable prediction for joining state)

## 1.0.9

- Simple preloader #61
- Simple mesh sprites #41
- Generic mesh virtual pads #64
- Improve Clear pass #62
- Improve Fog pass #63
- `[Escape]` key to navigate back #65
- Fade blood and footprints after some time #68
- Fix `step` sound playing interval

## 1.0.8

- Kid-mode blood #51
- Settings menu #48
- Dev menu: click 4 times on the main Game Logo
- Fix game camera smoothness
- Fix brightness for bone particles
- Weapon affect movement velocity #50
- Add Crosshair reload / empty state alterations
- Land particles and footprints #34
- Fix Weapon items overlaps the floor #42

## 1.0.7

- Offline Practice Mode #37
- Light below the Game Logo
- Cross-platform static monospace font for UI
- Splash Screen NPC start ammo
- Decrease minimum fire distance for Gun weapons for NPC
- Fix EMPTY / NO MAGS spam #44 (add reload / fire button down event) 
- Additional particles when bullet hits player shields
- Fix bullet bouncing on hit with damage #36 
- Keep logo while connecting #8
- Fix dancing nickname (prev Z used)

## 1.0.6

- Fix text particle spawn repeating
- Remove standard weapons drop from Barrels
- Limit max velocity for items drop
- Fix over camera shake effect for rail-gun
- Fix crosshair draw order
- Force ES2020 for client
- Fix resize
- Fix missing `speechSynthesis`
- Zoom out a little bit
- Splash screen on the game

## 1.0.5

### Graphics

- Font glyph combined with shadow
- Render more depth

### Gameplay

- NPC can't spawn with powerful weapon, make game start much easier

### UI

- Splash / Connecting screen improved 

## 1.0.4

- Change to OpenRelay STUN and TURN servers for WebRTC connection
- Change AI low-hp behaviour

## 1.0.3

- Change version label text to `v1.0.3 🏷`
- Fix server spam in `loaded` state before main menu
- Enable `POST` method to load `index.html`

## 1.0.2

### NPC
- NPC don't drop weapon until HP level above 5, then NPC start panic as before
- Slower NPC spawning
- Add Santa Claus NPC skin
- Add AI min-max shooting radius, searching for ammo

### General

- Add map themes #19
- Reset frags on death, add scores #18
- Add coin and diamond items for scores #18
- Add shield item and health-kit to add +2 hp
- Add ammo magazine item #11
- Add loot item lifetime, all items will be removed in 10 seconds #5
- Add weapon reloading mechanics #11
- Add [R] button - reload button to apply new ammo magazine #11
- Add [Q] button - swap secondary weapon slot #13
- Change [E] button to PICK weapon on the map or DROP your current weapon

### UI
- Add simple mini-map #10
- Add "change name" button #14 #23
- Improve Main Menu screen #23
- Add online users info in the Main Menu #23
- Add hyper-link to project available from game's version label #23
- Draw only enemy name label
- Add floating text when picking up items, or "out of ammo" case #24

## 1.0.1
- NPC are also fighting
- Simple text rendering
- HUD rendering
- Display client version
- Fix radial generation
- Add favicon
- Add FPS prefix
- Fix atlas pre-render, use consistent alphabetic baseline
- Display Name on top of character for real players
- Fix sound effects panorama (world scale issue)


base64sfxr.js
// http://stackoverflow.com/a/16001019

import {writeFileSync} from "fs";

function numberToFloat(bytes) {
    const sign = (bytes & 0x80000000) ? -1 : 1;
    let exponent = ((bytes >> 23) & 0xFF) - 127;
    let significand = (bytes & ~(-1 << 23));

    if (exponent === 128)
        return sign * ((significand) ? Number.NaN : Number.POSITIVE_INFINITY);

    if (exponent === -127) {
        if (significand === 0) {
            return sign * 0.0;
        }
        exponent = -126;
        significand /= (1 << 22);
    } else significand = (significand | (1 << 23)) / (1 << 23);

    return sign * significand * (2 ** exponent);
}


const b58alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

function b58_decode(S, A) {
    var d = [], b = [], i, j, c, n;
    for (i in S) {
        j = 0, c = A.indexOf(S[i]);
        if (c < 0) return undefined;
        c || b.length ^ i ? i : b.push(0);
        while (j in d || c) {
            n = d[j];
            n = n ? n * 58 + c : c;
            c = n >> 8;
            d[j] = n % 256;
            j++
        }
    }
    while (j--) b.push(d[j]);
    return new Uint8Array(b)
}

function convert(code) {
    const data = b58_decode(code, b58alphabet);
    const wave = data[0];
    const params = new Float32Array(23);
    params[0] = wave;
    let pi = 1;
    for (let i = 1; i < data.length; i += 4) {
        const val = (data[i] | (data[i + 1] << 8) | (data[i + 2] << 16) | (data[i + 3] << 24));
        params[pi++] = numberToFloat(val);
    }
    return params;
}

function f2a(f) {
    const x = Math.round(f * 1e6) / 1e6;
    console.log(x);
    return x.toString();
    // const a = f.toString();
    // let b = f.toPrecision(6);
    // while (b.length > 3 && (b[b.length - 1] === "0" || b[b.length - 1] === ".")) {
    //     b = b.substring(0, b.length - 1);
    // }
    // return a.length < b.length ? a : b;
}

function sign(s) {
    return s >= 0 ? 1 : -1;
}

function prepare(ps) {
    const wave = 0;
    const p_env_attack = 1;
    const p_env_sustain = 2;
    const p_env_punch = 3;
    const p_env_decay = 4;
    const p_base_freq = 5;
    const p_freq_limit = 6;
    const p_freq_ramp = 7;
    const p_freq_dramp = 8;
    const p_vib_strength = 9;
    const p_vib_speed = 10;
    const p_arp_mod = 11;
    const p_arp_speed = 12;
    const p_duty = 13;
    const p_duty_ramp = 14;
    const p_repeat_speed = 15;
    const p_pha_offset = 16;
    const p_pha_ramp = 17;
    const p_lpf_freq = 18;
    const p_lpf_ramp = 19;
    const p_lpf_resonance = 20;
    const p_hpf_freq = 21;
    const p_hpf_ramp = 22;


    ps[p_base_freq] = 100 / (ps[p_base_freq] ** 2 + 0.001);
    ps[p_freq_limit] = sign(ps[p_freq_limit]) * (100 / (ps[p_freq_limit] ** 2 + 0.001));
    ps[p_freq_ramp] = 1 - (ps[p_freq_ramp] ** 3) * 0.01;
    ps[p_freq_dramp] = -(ps[p_freq_dramp] ** 3) * 0.000001;
    ps[p_duty] = 0.5 - ps[p_duty] * 0.5;
    ps[p_duty_ramp] = -ps[p_duty_ramp] * 0.00005;

// p_arp_mod
    if (ps[p_arp_mod] >= 0) {
        ps[p_arp_mod] = 1 - (ps[p_arp_mod] ** 2) * .9;
    } else {
        ps[p_arp_mod] = 1 + (ps[p_arp_mod] ** 2) * 10;
    }

    if (ps[p_arp_speed] === 1) {
        ps[p_arp_speed] = 0;
    } else {
        ps[p_arp_speed] = (((1 - ps[p_arp_speed]) ** 2) * 20000 + 32) | 0;
    }

    ps[p_lpf_freq] = (ps[p_lpf_freq] ** 3) * 0.1;
    ps[p_lpf_ramp] = 1 + ps[p_lpf_ramp] * 0.0001;
    ps[p_lpf_resonance] = 5 / (1 + (ps[p_lpf_resonance] ** 2) * 20) * (0.01 + ps[p_lpf_freq]);
    if (ps[p_lpf_resonance] > 0.8) {
        ps[p_lpf_resonance] = 0.8;
    }

    ps[p_hpf_freq] = (ps[p_hpf_freq] ** 2) * 0.1;
    ps[p_hpf_ramp] = 1 + ps[p_hpf_ramp] * 0.0003;

// Vibrato
    ps[p_vib_speed] = (ps[p_vib_speed] ** 2) * 0.01;
    ps[p_vib_strength] = ps[p_vib_strength] * 0.5;

    ps[p_env_attack] = ((ps[p_env_attack] ** 2) * 100000) | 0;
    ps[p_env_sustain] = ((ps[p_env_sustain] ** 2) * 100000) | 0;
    ps[p_env_decay] = ((ps[p_env_decay] ** 2) * 100000) | 0;

    ps[p_pha_offset] = sign(ps[p_pha_offset]) * (ps[p_pha_offset] ** 2) * 1020;
    ps[p_pha_ramp] = sign(ps[p_pha_ramp]) * (ps[p_pha_ramp] ** 2);

    if (ps[p_repeat_speed] !== 0) {
        ps[p_repeat_speed] = (((1 - ps[p_repeat_speed]) ** 2) * 20000 + 32) | 0;
    }

    return ps;
}

function dumpProps(ps) {
    const aa = [];
    for(const x of ps) {
        aa.push(f2a(x));
    }
    return aa.join(",");
    //return ps.map(x => f2a(x)).join(",");
}

const sounds = [
    {
        name: ["blip", "heal", "med", "pick"],
        code: "34T6PktZ4axhapFFPhA7twnABw1FyGConUjS8Cjshjym6iNCamakVukZ6reS863897Kae2Bp3geSHwpXmPdFQD5nRUNYSEi1LrMcgRQQkizcA1oejht1thzxN",
    },
    {
        name: ["shoot"],
        code: "7BMHBGCZFc5aXpRKETBQrCjuxUuwoF1F9ovqMuZrUQGBZoaxzcfaju2BmsHwoKoDn94PoU8TXMZbj3a1vHybBzAjLAg23LTxxKZ5Fg2hm8vkquRVUSZWcNdHy",
    },
    {
        name: ["hurt"],
        code: "7BMHBGLMAznxzJ78hkrpuCXM31jegR1XjMXwBKbBZUF2zxQQA9g8gijGoRwDQZGYXhV7kgdDVvWQZKFKgXNrJRYjs8aLhSZsqe3ozazwDoPWhAXU43yKwuEo1",
    },
    {
        name: ["death"],
        code: "7BMHBGGKpyvRwcXtGttesStu6m23jzcDvcG8bnqS4SgPBx5554BK1puNbNfMwRpdTKMycqqXeZv8VSkTVvqPpqAsxZ5Krr3ABqC7DMjBSM1fH9StqLonjPZ6o",
    },
    {
        name:["step"],
        code: "7xZZDfZUkmERHMZGxWZxTXMdzAy7KnWbSCg1L13bqeCPvUfXgDPFGnMgmUMunc92GcQfsduC7fd8AiaBBJkLxfPZRradR6M5sUw6L6f8eNiP8cktDj7frKQLM",
    },
    {
        name: ["hit"],
        code: "11111LdfTv5fuPjEPZxtWMy1KMWu8ftYhofNkw9BguFGjEtUk1kNM3RJnizPEyfhNfUwmmJZUYTNHEU8VAUKCmnyMgmYvW5UVkBL92Cmh2Hvrbor8sbFFdH"
    }
];

const enumerations = [];
for (let i = 0; i < sounds.length; ++i) {
    const sound = sounds[i];
    const ps = convert(sound.code);
    prepare(ps);
    sound.props = ps;
    sound.i = i;
    for (const name of sound.name) {
        enumerations.push(name + " = " + i + ",");
    }
}

let code = `import {createAudioBuffer} from "../audio/sfxr";

export const enum Snd {
  ${enumerations.join("\n  ")}
}

export const snd:AudioBuffer[] = [];

export function loadSounds() {
  snd.push(
    ${sounds.map(x => `createAudioBuffer([${dumpProps(x.props)}]),`).join("\n    ")}
  );
}
`;

const floats = new Float32Array(1024 * 64);
let ptr = 0;

for(const sound of sounds) {
    for(const prop of sound.props) {
        floats[ptr++] = prop;
    }
}

writeFileSync("public/sfx.bin", floats.subarray(0, ptr));
writeFileSync("src/lib/assets/sfx.ts", code, "utf8");

// shoot
// const r = convert("7BMHBGCZFc5aXpRKETBQrCjuxUuwoF1F9ovqMuZrUQGBZoaxzcfaju2BmsHwoKoDn94PoU8TXMZbj3a1vHybBzAjLAg23LTxxKZ5Fg2hm8vkquRVUSZWcNdHy");
// jump
// const r = convert("34T6PktZ4axhapFFPhA7twnABw1FyGConUjS8Cjshjym6iNCamakVukZ6reS863897Kae2Bp3geSHwpXmPdFQD5nRUNYSEi1LrMcgRQQkizcA1oejht1thzxN");
// shooo
// const r = convert("11111Eh1QLewCdFgvpScbFobMMBFcDb6ToxYYSNKnhVmf1jk33vrHuiey9dftNbkzyDfCN9LY6uV3ctAtpvNNWRQFGtb5PdYHJBpWk6j6VZCoXZR5uXA6oGB");
// const ps = [...r.values()];